Add Identity policy cache

joestringer opened the pull request:
Based on #7542. See the last four commits for the changes proposed in this PR.

This PR adds a new package to manage the caching and lifecycle of
policies for identities. Long term this is intended to compute policy
for specific identities, handle policy or identity change events to
update the memoized form of these policies, and emit events to endpoints
to consume the newly computed policies.

The intended usage is that when we determine that an Identity is used
locally (hereafter Local Endpoint Identity, LEI [See #7616]), that LEI is inserted
into the distillery to notify it that SelectorPolicy should be resolved and
cached for that LEI. Policy changes should trigger a call to Update() the
policy for LEIs.

For the moment to keep the changes incremental, the endpoint triggers
the Update() from policy recalculation logic. Future work will look to
refactor the model around resolution and management of SelectorPolicy
away from the endpoint, but that is out-of-scope for this PR.

To incrementally introduce this model, the implementation currently only
handles generating SelectorPolicy objects and caching
them, then defers the generation of the EndpointPolicy (via Distill())
to users of the cache. Future commits are expected to extend this to
cache EndpointPolicy objects and track changes to global identities for
calculating differences in connectivity policy for peers via other
upcoming work (https://github.com/cilium/cilium/pull/7539).

This PR also adds some naive sharing logic within this identity policy cache
which uses the policy repository revision number to avoid recalculating
SelectorPolicy for an endpoint if it was already calculated for another
endpoint. As described in the comment, this is an optimistic attempt that
will work best in scenarios where new endpoints are created for an
existing Identity, but may potentially not work if regeneration for two
endpoints with the same Identity are triggered concurrently. No attempt is
made to keep SelectorPolicy alive beyond the lifetime of endpoints for now,
future work should configure a grace period to allow SelectorPolicy to be
reused if an endpoint is deleted and immediately readded on the same
node.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7617)
<!-- Reviewable:end -->

joestringer committed:
policy: Add distillery package

This commit adds a new package to manage the caching and lifecycle of
policies for identities. Long term this is intended to compute policy
for specific identities, handle policy or identity change events to
update the memoized form of these policies, and emit events to endpoints
to consume the newly computed policies.

The intended usage is that when we determine that an Identity is used
locally (hereafter Local Endpoint Identity, LEI), that LEI is inserted
into the distillery to notify it that SelectorPolicy should be cached
for that LEI. Policy changes should trigger a call to Update() the
policy for LEIs.

To incrementally introduce this model, the implementation currently only
handles generating selector-based SelectorPolicy objects and caching
them, then defers the generation of the EndpointPolicy (via Distill())
to users of the cache. Future commits are expected to extend this to
cache EndpoinytPolicy objects and track changes to global identities for
calculating differences in connectivity policy for peers via other
upcoming work (https://github.com/cilium/cilium/pull/7539).

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
endpoint: Consume policy from the distillery

Make use of the newly introduced policy distillery to cache and re-use
policies that were computed for other endpoints with the same Identity.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
testutils: Implement TestEndpoint.GetSecurityIdentity()

Allow the testutils.TestEndpoint use a real Identity backing and support
configuring the identity for more fleshed-out mocking of the endpoint
implementation.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
distillery: Skip policy resolution for same revision

When the distillery computes IdentityPolicy, first check whether this
expensive repository iteration would be a no-op and skip it if so.

This attempt is best-effort at this stage, as two endpoints may
currently race to this policy resolution step, both check the revision
and find that the policy is out-of-date, compute the identity policy,
then attempt to set it. In this case, only bump the policy if the
revision would be bumped.

Signed-off-by: Joe Stringer <joe@cilium.io>

jrajahalme commented on pkg/policy/distillery/distillery.go:
```diff
@@ -0,0 +1,184 @@
+// policyCache represents a cache of resolved policies for identities.
+type policyCache struct {
+	lock.Mutex
+	policies map[*identityPkg.Identity]*cachedSelectorPolicy
```
Does it matter that two Identity instances could describe the same (numeric) identity?

joestringer replied:
(As discussed oflfine) Yes.

 I'll look at switching this over to `NumericIdentity`.

jrajahalme commented on pkg/policy/distillery/distillery.go:
```diff
@@ -0,0 +1,184 @@
+	return changed, nil
+}
+
+// uldateSelectorPolicy resolves the policy for the security identity of the
```
"uldate..." -> "update..."

jrajahalme commented on pkg/policy/distillery/distillery.go:
```diff
@@ -0,0 +1,184 @@
+
+// Endpoint represents a user of an SelectorPolicy. It is used for managing
+// the lifetime of the cached policy.
+type Endpoint interface {
```
IMO "Owner" would be better name for this than "Endpoint", especially when this is NOT `endpoint.Endpoint`

joestringer replied:
It is currently, though. We don't want it to be longer term, but right now it is:

```
$ git grep "distillery\.[Upsert|Remove]"                                                                                                                                                                                                       
pkg/endpoint/endpoint.go:               if err := distillery.Remove(e); err != nil {
pkg/endpoint/policy.go:         e.SelectorPolicy = distillery.Upsert(e)
pkg/endpoint/policy.go: if err := distillery.UpdatePolicy(repo, e); err != nil {
```

joestringer replied:
By the way I'd like to trim this back from using `policy.PolicyOwner` to something simpler, but this is what's required for now as an incremental step.

joestringer replied:
I think I can add an incremental patch on top of this to switch it over already, actually. Let me try that.

joestringer replied:
Ah, so this is how the reference-counting is done currently. Basically the same SelectorPolicy can be found by two different endpoints, then the endpoint pointer is added to the list of `users`, which allows the distillery to avoid deleting a SelectorPolicy if it's still in use.

If we switch this over to use the identitymanager then it will no longer be necessary. However this will also require rearranging the lifecycle of the SelectorPolicy cache. I'd like to handle that in a separate PR.

jrajahalme reviewed (approved):
Small nits only.

joestringer committed:
distillery: Fix cardinality of cachedSelectorPolicy

Ensure 1:1 cardinality between identity -> cachedPolicy. Previously
there could be an issue where the same numeric identity had multiple
*Identity pointers, and the policies would not be shared between them.
This fixes it by indexing the policyCache by the numeric identity
instead.

Signed-off-by: Joe Stringer <joe@cilium.io>

aanm commented on pkg/policy/distillery/distillery.go:
```diff
@@ -0,0 +1,155 @@
+
+	cache.Lock()
+	defer cache.Unlock()
+	cip.setPolicyLocked(identityPolicy)
```
[1]

aanm commented on pkg/policy/distillery/distillery.go:
```diff
@@ -0,0 +1,155 @@
+	if len(cip.users) == 0 {
+		// TODO: Add deferred removal window in case the SelectorPolicy
+		//       may be re-used some time soon?
+		delete(cache.policies, identity)
```
[1]

aanm commented on pkg/policy/distillery/policy.go:
```diff
@@ -0,0 +1,59 @@
+)
+
+type cachedSelectorPolicy struct {
+	users  map[Endpoint]struct{}
```
This map is not being protected by any mutex, it's confusing to understand that this should be protected by the mutex in `[1]`

joestringer replied:
I'll add a comment, hopefully this makes it a bit clearer.

aanm commented on pkg/policy/distillery/policy.go:
```diff
@@ -0,0 +1,59 @@
+
+// setPolicyLocked updates the reference to the SelectorPolicy that is cached.
+func (cip *cachedSelectorPolicy) setPolicyLocked(policy *policy.SelectorPolicy) {
+	atomic.StorePointer(&cip.policy, unsafe.Pointer(policy))
```
if you are using atomic.Store/Load you don't need to hold any lock. This is a little bit confusing as well because `cip.policy` could simply be a `*policy.SelectorPolicy` which would be wrapped around a mutex that would save concurrent access of `users` and `policy` fields

joestringer replied:
The lock isn't for the pointer storage, it's for keeping the `cip.policy` and `cip.revision` in sync. This might not be apparent until the later patches in the series.

The design of this (which I recognize is not documented, will fix) hands the `cachedSelectorPolicy` reference to an endpoint, which I want to be able to fetch the latest version of the policy without holding any locks (See the `Consume()` function). This is similar to to the approach used in #7539 , effectively "read and concurrent update"; the atomic provides a "publish" mechanism of an immutable copy of the most recently computed selector policy, and the policy calculation side can calculate and update the pointer without locking readers out.

joestringer replied:
I've added additional comments to try to highlight this.

joestringer replied:
And now I've implemented a change to streamline this.

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -288,6 +289,10 @@ type Endpoint struct {
 
+	// SelectorPolicy represents a reference to the shared SelectorPolicy
+	// for all endpoints that have the same Identity.
+	SelectorPolicy distillery.SelectorPolicy `json:"-"`
```
does it need to be an exported field?

joestringer replied:
Probably not, good catch.

joestringer committed:
policy: Add and use Revision in SelectorPolicy

A generated SelectorPolicy is created with a particular revision of the
repository, so it makes sense to track this information within the
SelectorPolicy itself. This allows the distillery to be simplified,
because now there doesn't need to be locking to ensure that the
cachedSelectorPolicy's policy and revision fields are in sync.

Signed-off-by: Joe Stringer <joe@cilium.io>

jrajahalme commented on pkg/policy/distillery/policy.go:
```diff
@@ -0,0 +1,72 @@
+// 'policy' field). It is always nested directly in the owning policyCache,
+// and is protected against concurrent writes via the policyCache mutex.
+//
+// 'policy' and 'revesion' are only consistent if queried while holding a lock.
```
->'revision', but then the 'revision' is no longer here, so needs an update?

joestringer replied:
I'll tidy it up in a follow-up PR.
